worlds/sc2wol/PoolFilter.py

```diff
@@ -18,46 +20,59 @@
+        MissionPools.EASY: [mission.mission_name for mission in get_missions_by_pool(MissionPools.EASY)],
+        MissionPools.MEDIUM: [mission.mission_name for mission in get_missions_by_pool(MissionPools.MEDIUM)],
+        MissionPools.HARD: [mission.mission_name for mission in get_missions_by_pool(MissionPools.HARD)],
+        MissionPools.VERY_HARD: [mission.mission_name for mission in get_missions_by_pool(MissionPools.VERY_HARD)],
         MissionPools.FINAL: []
     }
     if mission_order_type == MissionOrder.option_vanilla:
         # Vanilla uses the entire mission pool
-        mission_pools[MissionPools.FINAL] = ['All-In']
+        goal_priorities: Dict[SC2Campaign, SC2CampaignGoalPriority] = {campaign: get_campaign_goal_priority(campaign) for campaign in enabled_campaigns}
+        goal_level = max(goal_priorities.values())
+        candidate_campaigns = [campaign for campaign, goal_priority in goal_priorities.items() if goal_priority == goal_level]
+        goal_campaign = multiworld.random.choice(candidate_campaigns)
+        mission_pools[MissionPools.FINAL] = [campaign_final_mission_locations[goal_campaign].mission.mission_name]
+        remove_final_mission_from_other_pools(mission_pools)
         return mission_pools
     # Omitting No-Build missions if not shuffling no-build
     if not shuffle_no_build:
-        excluded_missions = excluded_missions.union(no_build_regions_list)
-    # Omitting Protoss missions if not shuffling protoss
-    if not shuffle_protoss:
-        excluded_missions = excluded_missions.union(PROTOSS_REGIONS)
-    # Replacing All-In with alternate ending depending on option
-    if final_map == FinalMap.option_random_hard:
-        final_mission = multiworld.random.choice([mission for mission in alt_final_mission_locations.keys() if mission not in excluded_missions])
-        excluded_missions.add(final_mission)
+        excluded_missions = excluded_missions.union([mission.mission_name for mission in get_no_build_missions()])
+    # Omitting missions not in enabled campaigns
+    for campaign in disabled_campaigns:
+        excluded_missions = excluded_missions.union([mission.mission_name for mission in get_campaign_missions(campaign)])
+
+    # Finding the goal map
+    goal_priorities = {campaign: get_campaign_goal_priority(campaign, excluded_missions) for campaign in enabled_campaigns}
+    goal_level = max(goal_priorities.values())
+    candidate_campaigns = [campaign for campaign, goal_priority in goal_priorities.items() if goal_priority == goal_level]
+    goal_campaign = multiworld.random.choice(candidate_campaigns)
+    primary_goal = campaign_final_mission_locations[goal_campaign]
+    if primary_goal is None or primary_goal.mission.mission_name in excluded_missions:
+        # No primary goal or its mission is excluded
+        candidate_missions = list(campaign_alt_final_mission_locations[goal_campaign].keys())
+        goal_mission = multiworld.random.choice(candidate_missions).mission_name
```

Comment:
This should probably raise a more descriptive exception than `IndexError` if `candidate_missions` is empty, eg. "Couldn't place final mission, please exclude fewer missions".

Reply:
Updated